packages/kas/src/lexer.ts

```diff
@@ -0,0 +1,192 @@
+    ["sinh|cosh|tanh", (value) => ({kind: "TRIG", value})],
+    ["csch|sech|coth", (value) => ({kind: "TRIG", value})],
+    ["\\\\sin", {kind: "TRIG", value: "sin"}],
+    ["\\\\cos", {kind: "TRIG", value: "cos"}],
+    ["\\\\tan", {kind: "TRIG", value: "tan"}],
+    ["\\\\csc", {kind: "TRIG", value: "csc"}],
+    ["\\\\sec", {kind: "TRIG", value: "sec"}],
+    ["\\\\cot", {kind: "TRIG", value: "cot"}],
+    ["\\\\arcsin", {kind: "TRIG", value: "arcsin"}], // should this be TRIGINV
+    ["\\\\arccos", {kind: "TRIG", value: "arccos"}], // should this be TRIGINV
+    ["\\\\arctan", {kind: "TRIG", value: "arctan"}], // should this be TRIGINV
+    ["\\\\arccsc", {kind: "TRIG", value: "arccsc"}], // should this be TRIGINV
+    ["\\\\arcsec", {kind: "TRIG", value: "arcsec"}], // should this be TRIGINV
+    ["\\\\arccot", {kind: "TRIG", value: "arccot"}], // should this be TRIGINV
+    ["arcsin|arccos|arctan", (value) => ({kind: "TRIGINV", value})],
+    ["arccsc|arcsec|arccot", (value) => ({kind: "TRIGINV", value})],
+    ["\\\\sinh", {kind: "TRIG", value: "sinh"}],
+    ["\\\\cosh", {kind: "TRIG", value: "cosh"}],
+    ["\\\\tanh", {kind: "TRIG", value: "tanh"}],
+    ["\\\\csch", {kind: "TRIG", value: "csch"}],
+    ["\\\\sech", {kind: "TRIG", value: "sech"}],
+    ["\\\\coth", {kind: "TRIG", value: "coth"}],
+    ["pi", {kind: "CONST", value: "pi"}],
+    ["\u03C0", {kind: "CONST", value: "pi"}], // pi
+    ["\\\\pi", {kind: "CONST", value: "pi"}],
+    ["theta", {kind: "VAR", value: "theta"}],
+    ["\u03B8", {kind: "VAR", value: "theta"}], // theta
+    ["\\\\theta", {kind: "VAR", value: "theta"}],
+    ["phi", {kind: "VAR", value: "phi"}],
+    ["\u03C6", {kind: "VAR", value: "phi"}], // phi
+    ["\\\\phi", {kind: "VAR", value: "phi"}],
+    ["[a-zA-Z]", (value) => ({kind: "VAR", value})], // TODO: post-process
+    [".", {kind: "INVALID"}],
+];
+
+type PreparedRule = [RegExp, Token | ((match: string) => Token)];
+
+const preparedRules: PreparedRule[] = rules.map((rule) => {
+    const regex = new RegExp("^(?:" + rule[0] + ")");
+    return [regex, rule[1]];
+});
+
+// TODO:
+// - try each of the rules in order
+// - pick the longest one and return token
+// - update the input string so that we match the next token
+
+const next = (
+    input: string,
+    preparedRules: PreparedRule[],
+): [Token, string] => {
+    let currentMatch: string = "";
+    let token: Token = {kind: "INVALID"};
+
+    for (const rule of preparedRules) {
+        const match = input.match(rule[0]);
+        if (match && match[0].length > currentMatch.length) {
+            currentMatch = match[0];
+            if (typeof rule[1] === "function") {
+                token = rule[1](currentMatch);
+            } else {
+                token = rule[1];
+            }
+        }
+    }
+
+    if (currentMatch === "") {
+        throw new Error(`No match for ${input}`);
+    }
+
+    return [token, input.slice(currentMatch.length)];
```

Comment:
I don't think it would that much additional complexity.  I did it this way because that's what jison-lex does, see https://github.com/zaach/jison-lex/blob/master/regexp-lexer.js#L330.  I'll try keeping track of the current index instead, but to do so I think restructuring the code to use a class for the lexer will help.